### ISSUE TITLE
I'm at a loss, can anyone help

### ISSUE THREAD

**Issue by Dunravin** [uglyDwarf/linuxtrack#206]
Hello.

I am an Arch newbie, I have a TrackIR 5 device I have been attempting to get working on linux.

So far I've tried installing linuxtrack as per the instructions but I fail when I get to testing for dependencies as all the dependencies are old and I don't know how to deal with that at my current level or experience.

I have also tried to build linuxtrack from source from this repo and several of the forks but once again I'm getting very lost and have failed.
I've also tried adding linuxtrack from the Aur repositories, there's two versions both out of date and neither work.

I was chatting with some guys on the LUG Star Citizen discord and one suggested Opentrack might be a solution. I can get Opentrack to install but from there I have no idea how to add the necessary things to access the TrackIR hardware as a source or get any output.

I make out from the linuxtrack info that something is needed from TrackIR which is extracted with Wine.

So I guess the main question is persevere with linuxtrack which I know will address the hardware or go the opentrack route which is an unknown as to whether it will accept TrackIR 5 camera input.

I am completely lost, if anyone would be willing to help me I'd be eternally grateful. Some hand holding may be rquired.

Many thanks D.

**Comment by Peace-art-cloud** [comment]
You're not the only one, pal. I am at my wit's end trying to figure this stuff out to get my TrackIR working. Nothing seems straight forward. I follow one instruction but then it don't work, I try another that works a bit but then don't. then there's no explanation on how to bridge a gap. I've sat here for hours trying to figure this out just to get my Track Ir 5 working. the instructions seem to have gaps, like trying to cross a bridge but then there is no bridge.

It feels like a test: If you're stupid no Linux for you and they just watch us to see how stupid we are. "Use that bridge, kid, yea I know you don't see a bridge, you gotta figure out how to build that bridge to get across. There's some wood over there."

Yes I  am total noob too, and the issue is no doubt on our end, but experienced Linux users doesn't seem to understand it may not be obviously clear to us noobies.

Sorry, not meaning to insult anyone,  just frustrated,

**Comment by immakyle** [comment]
@Dunravin @Peace-art-cloud I was in your shoes not long ago, still kind of am. I started with linuxtrack and more recently migrated to opentrack because of active development. From what I've gathered, my advice is this...

1. Depending on how you use TrackIR (with trackclip or trackclip pro) will determine whether you can use opentrack with NaturalPoint hardware. If, like me, you only have the trackclip that attaches to the brim of a baseball cap you will not be able to use TrackIR with opentrack. From my understanding, you can use the trackclip pro for use with a webcam in opentrack. But for legal reasons opentrack doesn't have compatibility with TrackIR sensor/hardware. Need webcam.

2. Linuxtrack is preferred for this reason. It can use the TrackIR sensor which is vastly superior. Problem here is outdated software (which is not really a problem on Arch, I'll explain in a bit), and no built-in Steam/Proton support (again, it's old software). So even when you get linuxtrack up and running it won't communicate with your Steam titles, other than X-plane because linuxtrack comes with an X-plane plugin that handles the communication. If I knew how to write other plugins I could solve this problem for myself, but I can't. Judging by this post, neither can you.

That said, if you still want to try linuxtrack...do yourself a favor and use the universal build and wiki instructions [here](https://github.com/uglyDwarf/linuxtrack/wiki/Universal-Linuxtrack-Package). Dependencies are listed. When you were testing for dependencies, the terminal should have spit out what you were missing. Typically I search for dependencies by running a simple google search like, "libqtgui4 arch linux". Ideally a package pops up, or can be deduced where to find after a little reading. 

Thankfully the AUR coverage is vast and a lot of old stuff can be found. Even easier for linuxtrack because it has two outdated linuxtrack packages! [Here](https://aur.archlinux.org/packages/linuxtrack-bin) is one you already saw. Important part is it has the dependencies listed below, which the wiki above noted. In my experience, you don't need them all. For example cwiid is for use with wiimote function/capability of linuxtrack. I don't have a wiimote so I don't need it, and more importantly linuxtrack works fine without it. I digress... follow the wiki for needed packages. Get them from arch official repos when possible, otherwise popular/upvoted AUR packages will do. Anyway...that should get linuxtrack going. 

Hoping for an updated universal package soon because TrackIR hardware is best.

3. If, like me, you can't find a way to get other games working with linuxtrack (excluding X-plane, works great there btw)...you're onto the next portion of this mystical journey. Opentrack! If you thought linuxtrack instructions were bad, think again! You can get opentrack to compile just fine, but again, it is the communication with Steam/Proton I can't figure out when trying to use the native linux version of opentrack. Here is the [solution](https://github.com/markx86/opentrack-launcher). That script runs the Windows portable version of opentrack through Proton. Follow instructions for non-root option and it will execute when you start your game. After configuring and deciding which inputs/outputs you need (depending on your equipment), you will have headtracking working in-game. 

Remember, I don't have trackclip pro. I bought a cheap webcam and use neuralnet as input and freetrack 2.0 as output. Tracks my face. It's a compromise until linuxtrack is updated or something better comes along. 


I have regurgitated everything I know about this process. Hopefully this helps you both. Please give these github projects stars and support them. Despite lack of documentation... they are, indeed, all we have.

**Comment by fwfa123** [comment]
Compiled from several threads on the issue tracker...here are the steps I took to get it running on my distro.
All the highlighted code below is to be copied and entered into your own linux terminal

`mkdir -p $HOME/Code`

`cd $HOME/Code/`

`sudo dpkg --add-architecture i386`

`cat /etc/os-release`
Make a note of your distribution name:
Look for the line with either UBUNTU_CODENAME or VERSION_CODENAME. If both are present, use the name after UBUNTU_CODENAME.
*****info for the above if you need to know https://gitlab.winehq.org/wine/wine/-/wikis/Debian-Ubuntu

`sudo mkdir -pm755 /etc/apt/keyrings`

`sudo wget -O /etc/apt/keyrings/winehq-archive.key https://dl.winehq.org/wine-builds/winehq.key`

My distro is MX Linux 23.3 =  Bookworm Debian 12
`sudo wget -NP /etc/apt/sources.list.d/ https://dl.winehq.org/wine-builds/debian/dists/bookworm/winehq-bookworm.sources`

`sudo apt update`

'sudo apt install --install-recommends -y winehq-devel wine-devel-dev git checkinstall build-essential automake bison gawk flex gcc-multilib g++-multilib gobjc gobjc++ libmxml-dev libssl-dev libusb-1.0-0-dev libcwiid-dev libopencv-dev libv4l-dev libqt5opengl5-dev libqt5webkit5-dev libqt5x11extras5-dev qttools5-dev qttools5-dev-tools nsis`
*****For me in the list above the libopencv-dev was a deal killer but my distro already had some opencv libraries so I moved on as I believe it has to do with face tracking support and I would not be using that with my trackir. I also added qt6-dev and qt6-dev-tools to the list as some of the linuxtrack repositories are moving towards qt6 and I added it just in case.

Choose a repo to try...I ran through the list below one repo at a time...downloading and then working through each of the steps shown below all the way until the last step where you run the gui _"$HOME/linuxtrack/bin/ltr_gui"_ if at that point I received an error I backed up to this point and downloaded and tried the next repo. I did this 4 times in my tries and the 4th one worked.
git clone https://github.com/exuvo/linuxtrack.git  Failed
git clone https://github.com/rfvizarra/linuxtrack  Failed
git clone https://github.com/BarronKane/linuxtrack-qt6 Failed
`git clone https://github.com/RavenX8/linuxtrack`

`cd linuxtrack/`

******DEPENDING on which linuxtrack repo you choose you may need to add the following
modify the file ltr_srv_master.cpp and adding #include <pthread.h> as first line.  The file is located in the $Home/Code/linuxtrack/src/ folder if you used the steps above
******Once modified or confirmed continue with the below


`aclocal`

`autoupdate`

`autoconf`

`automake`

`./configure --prefix=$HOME/linuxtrack --with-wine-libs=-L/opt/wine-devel/bin`

`make -j`

`sudo make install`

`sudo cp $HOME/linuxtrack/share/linuxtrack/99-TIR.rules /lib/udev/rules.d`

reboot

Plug in your Trackir device and enter this in your terminal
`$HOME/linuxtrack/bin/ltr_gui`
If you have your device plugged in it should say it found device # and launch the gui

**Comment by immakyle** [comment]
How are you able to have games recognize the head tracking? On steam (through proton), or otherwise? I've only had luck with the built-in X-plane plugin.

**Comment by fwfa123** [comment]
@immakyle Sorry...nothing yet.  I just got this working but I have not had luck with wine or proton at this time.  I will update if I get it working in a game under one of these systems

**Comment by fwfa123** [comment]
@immakyle Some good news, I was able to get Linuxtrack up and running in Falcon 4 using the builtin #HOME/linuxtrack/share/linuxtrack/linuxtrack-wine.exe included when you install Linuxtrack.

I had a lot of back and forth steps trying to figure it out so hopefully the steps below help out..

**Install the Trackir firmware**, if you have not already. Steps to complete this are shown below.

Launch the ltr gui
`$HOME/linuxtrack/bin/ltr_gui`

Select the Misc. tab
Click on the Install TrackIR firmware

Here you can download and allow it to install it for you but I went with my install of trackir in wine. I had installed it earlier in wine to see if I could get that to work...nope.  So for me I selected "Extract from unpacked"  most people will probably use "Extract from installer".
![tir1](https://github.com/user-attachments/assets/1b8b2179-5725-4d09-87cd-2990f6a7dfc1)

**Install the wine bridge app**

Now I was not able to use the button on this page to "Install Linuxtrack-Wine support..." so I had to install the support manually.
At this point you should use wine to run this program "#HOME/linuxtrack/share/linuxtrack/linuxtrack-wine.exe"  inside of the wine prefix where you have the game installed that you want to use with Linuxtrack.  For me this is simple as I use Lutris to run and manage all of my windows games.  

In Lutris you select your game from inside the app(Lutris) and then at the bottom of the apps(Lutris) page click the right arrow and select "Run EXE Inside Wine Prefix"
![TIR2](https://github.com/user-attachments/assets/b9dce2d9-0c50-41f0-88b4-3ba13bb70c44)
Select the linuxtrack-wine.exe from the directory #HOME/linuxtrack/share/linuxtrack/. Let the installer run and select okay or finish.

**Place the missing TIRViews.dll.**
 Now here is where I did have to have trackir installed in either a windows install, VM windows install, or a wine install of it. For some reason when the firmware is extracted this dll was not being copied over so I had to find a copy of it manually and then track the simlink back to where it was pointing from my wine prefix.

I found the dll in my trackir install directory "C:\Program Files (x86)\NaturalPoint\TrackIR5\" again this could be in your trackir install in wine as i had done as a test earlier or in a windows install.
Copy the TIRViews.dll and paste it in the #HOME/.config/linuxtrack/tir_firmware/ folder. You may have to open this folder as root to paste but I was able to open and paste without being root.

That was it...for me I launched my Falcon4 in Lutris and my trackir turned on automatically and was ready to go...  I had to check use track ir in my controller advanced settings under Falcon4 BMS.  

If you need to update the game profile list, as some need this profile to run, use
`~/linuxtrack/bin/ltr_extractor --update`

I hope this helps some one and I plan on testing proton in steam soon

**Comment by immakyle** [comment]
> @immakyle Some good news, I was able to get Linuxtrack up and running in Falcon 4 using the builtin #HOME/linuxtrack/share/linuxtrack/linuxtrack-wine.exe included when you install Linuxtrack.
> 
> I had a lot of back and forth steps trying to figure it out so hopefully the steps below help out..

Thank you so much for the detailed walkthrough, and I appreciate you getting back to me. These steps are very similar to another post I had found during my initial experimentation with getting linuxtrack working. IIRC they were using protontricks to install the wine-bridge inside game prefix, which seemed to work as expected, but perhaps it was also missing TIRViews.dll. I'll have to give this a try with Lutris over the weekend. Thanks again, this helps a lot!

**Comment by fwfa123** [comment]
@immakyle You are welcome, I will have to give protontricks a whirl to see what happens.  

One other quick thing I wanted to post that may be helpful was another little jewel that @uglyDwarf built in to linuxtrack
**Linuxtrack to antimicrox to Joystick/Mouse/Keyboard**

Some uses where I apply it is…  Janes Longbow2 running in PCEM. Janes Longbow2 has a 3D cockpit mode where the view can be controlled by joystick or the number keypad. I use antimicrox to translate my linuxtrack server output to the Number keypad. You can adjust the speed of your head movement in the longbow ca.ini file.

Another example is running Mechwarrior 2 in dosbox. Mechwarrior 2 uses your mouse to control your torso view and targeting reticle. So I have antimicrox pass the Pitch and Yaw axis to the Mouse and then I can use head tracking in MECH2. 

If you do not have antimicrox you can probably install it through your distros software repository. Most of them have this built in to theirs so it is a quick and easy install.  Info https://antimicrox.github.io/

Check to see if your user is in the plugdev group.
Run groups in the terminal
`groups`
If you do not see the plugdev group in the list add your user to that group
`sudo usermod -a -G groupname username`
Replace the username listed in the command with your user name

Setup the input with these terminal commands. Run one at a time

`modprobe uinput`
`sudo chgrp plugdev /dev/uinput`
`sudo chmod 660 /dev/uinput`

Launch linuxtrack before running the server below.
`~/linuxtrack/bin/ltr_gui`

Joystick emulation mode:

`~/linuxtrack/bin/ltr_pipe --output-file=/dev/uinput --format-uinput-abs`
Leave this terminal open

Launch antimicrox...or if you already had it open you should see that it now sees your Linuxtrack as a Joystick. The axis are seen by antimicrox as follows. 

    • Axis 1 = Yaw
    • Axis 2 = Pitch
    • Axis 3 = Roll
    • Axis 4 =  Move left/right
    • Axis 5 Move up/down
    • Axis 6 Move backwards/forwards
![Antimicrox1](https://github.com/user-attachments/assets/274d1875-f66e-4dd1-b8f9-3cb0b573a963)

From here you can start to edit what each axis does.
![Antimicrox2](https://github.com/user-attachments/assets/e308dcff-0a55-47dd-a8a9-d967119283a1)

**Comment by immakyle** [comment]
@fwfa123 I finally had an opportunity to try out your steps to get linuxtrack working... I honestly can't thank you enough. This is going to significantly improve the small bit of gaming I get to do these days. Also works with proton/steam using the same wine bridge method. 

I noticed NaturalPoint has been releasing new firmware in recent months after a long period of no releases (after version 5.4.2). Many have chimed in over the years on their forum requesting a driver that works on linux. Can't say they will do it, but I'm hopeful with the recent releases.


Things to note (for anyone else who comes across this thread):
Distro: Arch (qt4, qtwebkit, openssl-1.0 are available in AUR and nsis, mxml from official repos). Was all I needed to run universal package. Follow build instructions from earlier in this thread to build a working fork from source.

1. I needed to use both a working fork (for a functional wine bridge known as linuxtrack-wine.exe) and the universal package (for an error-free firmware install) to get linuxtrack running properly. Best forks as of the time of writing: RavenX8 and marengohue.

Firmware is installed in .config/linuxtrack. Both forks have issues with installing firmware error-free. "Extract from installer" button resulted in an extraction error, and "Extract from unpacked" worked using my wine prefix (.wine) but resulted in errors in gamedata.txt. I used latest available version of TrackIR firmware (5.5.3). I tested on Arma 3. While the sensor would turn on during game start... there was no input, signaling that wine bridge functioned but something was wrong with the firmware. Took a look at gamedata.txt and noticed "((null))" next to Arma 3 and various other titles. I don't know why this happens but the issue is present on both forks. So, I instead used the universal package to install firmware without problems. Although, even with the universal package...I couldn't use "Extract from installer". I needed to download TrackIR software from the [website](https://www.trackir.com/downloads/) and use wine to run it. That installs it to the default prefix (.wine). Then use "Extract from unpacked" on .wine directory.

2. I also needed to use the linuxtrack universal package to extract mfc42u.dll from VC6RedistSetup_deu.exe. That dll file is necessary and supposed to be attainable through the "Install Linuxtrack-Wine support..." button, which prompts to download Microsoft Visual C++ 6.0 Redistributable package. That button doesn't work on either fork. At times I've also had the download fail using the universal package. If the download fails the workaround for universal package is to search the download link in web browser manually. Once you download the file, then use "Extract from installer" on VC6RedistSetup_deu.exe. Worked for me and it placed mfc42u.dll in ~/.config/linuxtrack/tir_firmware.

3. Since the "Install Linuxtrack-Wine support..." button doesn't work, I needed to manually install the wine bridge/linuxtrack-wine.exe (from linuxtrack fork, not universal package) into the proton prefix. As stated earlier in the thread, you can use Lutris for this. You can also do it in your terminal by using "export WINEPREFIX=/home/_username_/.steam/steam/steamapps/compatdata/_app ID_/pfx". The app ID for your desired game can be found by right-clicking game title in steam library, properties, updates. Then use wine to run linuxtrack-wine.exe. A pop up will appear, click through it. When it's complete you'll see Linuxtrack in Program Files (x86) in the proton prefix.

4. Lastly, to update your gamelist.txt you can use ltr_extractor --update. Again, I needed to use the binary provided in the universal package for this otherwise certain titles would be listed as "((null))" and fail to receive input.

**Comment by fwfa123** [comment]
@immakyle This is some great info..I never was able to get around to the proton side of things. My test was going to be for ARMA3 when I got there.  But it appears you solved the issue for me before I got started.  Well documented, Thanks again!

**Comment by Dunravin** [comment]
@fwfa123 Sorry for the late reply. I have been rather busy but finally got around to tinkering with getting the TrackIR working again.

Thanks for all your input and help, it's really appreciated. 

I got through the steps to compile with the RavenX8 repo but ended up with no ltr-gui.
I suspect this is because I need qt installed. I have tried again with qt5 and qt6 installed from the arch repositories but no luck.

I am currently building qt4 from an AUR repository and will try to compile linuxtrack again once that's finished. qt3 threw errors so I pray that's not the version I need if in fact my issue is with qt not being present. (Still very wet behind the ears in the linux world)

[EDIT]
qt6-tools qt5-tools got me the gui. Not sure which, don't care. The gui works and TrackIR is tracking! 

I was still missing linuxtrack-wine.exe 

To get that I checked I had the wine development files

`whereis winegcc`
`whereis wineg++`

and install nsis from an Arch AUR repo as per /Code/linuxtrack/src/wine_bridge/README.wine

then edit the linuxtrack configure command

`./configure --prefix=$HOME/linuxtrack --with-wine-libs=-L/opt/wine-devel/bin --enable-wine-plugin`

and continue with the make as per your instructions.

I have yet to attempt to get tracking working in my steam games but I suspect `protontricks` will be the way to install the TrackIR stuff to the game prefixes. Will update this post when success.

**Comment by Dunravin** [comment]
Update -

`./ltr_extractor --update`  did connect and download the sig file but when I check the gamedata.txt file elite dangerous is ((null))

Had to manually copy in TIRView.dll and the other dll, that one I had to download because I could not get the GUI extractor to process VC6RedistSetup_deu.exe. I put them in the game folder where I believe they ought to be but no success getting tracking in Elite although it works fine in the ltr_gui.

**Comment by immakyle** [comment]
> Update -
> 
> `./ltr_extractor --update` did connect and download the sig file but when I check the gamedata.txt file elite dangerous is ((null))
> 
> Had to manually copy in TIRView.dll and the other dll, that one I had to download because I could not get the GUI extractor to process VC6RedistSetup_deu.exe. I put them in the game folder where I believe they ought to be but no success getting tracking in Elite although it works fine in the ltr_gui.

You're nearly there. The ((null)) entries are the issue and happens when you use the linuxtrack forks to update gamedata.txt. If you're still on Arch, I recommend gathering as many dependencies as possible (qt4, qtwebkit, openssl-1.0 are available in AUR and nsis, mxml from official repos) to run universal package for installing the firmware and updating games list. 

Either that, or since Elite is kind of an older title, redo the install and just don't use ltr_extractor --update. Look at the initial gamedata.txt and see if Elite is there. If so, should work. Only need to update the gamelist for new TrackIR supported titles, in which case you should get the universal package working anyway.

**Comment by fwfa123** [comment]
@immakyle  Glad to hear you are having some success and thank you for publishing your path to getting it running in Arch.
@Dunravin  Thank you for the continued responses as they are helping me also. 
One thing that has helped me when I could not get a particular game to run is that I will go to the wine environment of the game I am trying to run and under the linuxtrack install folder there are a couple of programs there that you can launch to test the linuxtrack interface without launching the game. Sorry, I am not in front of my PC wthat has linuxtrack, but I believe it is called Tester64.exe . I have found that once I launch that in the environment and it works. I go back to the game and it will then work.  Not sure why but it works.

Also, I have had NO luck running inside of proton (Steam). If you guys have any hints or a quick breakdown I would appreciate it.  I did install Arch to try and run that route but for some reason QT4 would not compile through AUR. Are you guys running straight Arch and building from the ground up or a Distro such as Manjaro?

**Comment by immakyle** [comment]
@fwfa123 I'm using straight Arch. 

I forgot to mention that I've also had issues with compiling qt4 from the latest 2 commits (pkgver 37 & 38). Last one that works for me is pkgver 36 [here](https://aur.archlinux.org/cgit/aur.git/commit/?h=qt4&id=5a6bb417c40abec944b564b7c39bae86f4b8884b). Up top you'll see that you can download the tar.gz. 

Extract it:
tar xvf Downloads/aur-5a6bb417c40abec944b564b7c39bae86f4b8884b.tar.gz -C Downloads/

Build & install:
cd Downloads/aur-5a6bb417c40abec944b564b7c39bae86f4b8884b
makepkg -si

Once you have qt4 and the other dependencies I've mentioned in previous post you'll be able to run the [universal package](https://github.com/uglyDwarf/linuxtrack/wiki/Universal-Linuxtrack-Package). In my experience you need to use it for everything except installing the wine bridge (linuxtrack-wine.exe), which you can do using a working fork. Below I've posted screenshots showing what appears in my directories after completing all the steps. Note the file paths and symlinks. The example directory in the screenshot "1250410" is the app ID for Microsoft Flight Simulator 2020 on Steam. Also to clarify, don't worry about the dll or exe files you see in .config/linuxtrack. Those are copies/backups of the files we're using. I don't want to lose them in case they become unavailable in the future. As long as the dll files are in tir_firmware and symlinked in the proton prefix, you're good.

I might be able to document my steps with a list of commands, one by one, but I need more time to write that out.

![Image](https://github.com/user-attachments/assets/5c0b7f11-19a5-416a-be9b-dc936bd1b0d3)
![Image](https://github.com/user-attachments/assets/f8d03aaa-2a80-49be-ad7c-1ee7e916edab)

**Comment by immakyle** [comment]
Afterwards you can edit /etc/pacman.conf. Uncomment "IgnorePkg =" and type in qt4. That way whenever you update your system packages, pacman will ignore trying to update qt4.

**Comment by fwfa123** [comment]
All good info, never have installed straight Arch before but it is time for me to put my big boy pants on a jump in to that end of the pool.  Plus I will drop back to proton and see about the info setup like you have shown. Thanks again for the reply

**Comment by Goosoe** [comment]
This thread helped me a lot to manage to make linuxtrack work with TrackIR 5 on arch 6.15.4 with Steam proton games. As a thank you I decided to write the biggest stumbles I came across and what I did to fix them in order to make this work. Maybe this can help someone in the future as well.

- First off, I grabbed the latest on master. Building it required me to include the pthread library (`#include <phterad.h>`) in a specific .cpp file, plus remove some deprecated values in a couple of switch case regarding the Wine library (DLL_WINE_PREATTACH) on other files.

- The current wine package in arch merged the 32 and 64 bit builds and moved to the WoW64 build (https://archlinux.org/news/transition-to-the-new-wow64-wine-and-wine-staging/). This means that we do not have the wine 32 bit binaries. This is a big problem since with the lack of them, we won't be able to build and compile the so desired linuxtrack-wine.exe. To fix this, at the moment, I switched to the wine-stable (https://aur.archlinux.org/packages/wine-stable) package which still has the 32-bit version of winegcc and wineg++ to build the linuxtrack-wine.exe.

- Afterwards, using the steps provided by @fwfa123  (https://github.com/uglyDwarf/linuxtrack/issues/206#issuecomment-2381328185) I had everything compiled properly.

- Now we can run $HOME/linuxtrack/bin/ltr_gui and see if things are working okay. I had to disable 3D Tracking Preview in the Misc. tab to be able to get the camera view without lag. Everything was set up nicely to work natively, but now the other problem: I wanted to play some games through Steam Proton, such as IL-2. For that, we had to install the bridge in the desired game prefix.

- To properly install everything in the game prefix, I used Protontricks (https://aur.archlinux.org/packages/protontricks).
I had to install some things in there. Initially I ran the command in the console `protontricks <GameId on steam> -q mfc42` to install various .dll's but we were aiming at this one specifically: mfc42u.dll.
Then I opened Protontricks directly on the game prefix and installed the latest TrackIR5 software, which I downloaded from TrackIR official website.
Finally I manually installed the previously built linuxtrack-wine.exe. 
The problems did not end here. Firstly, the IR camera would not turn on when the game was launched. The solution was found mentioned in a comment above: Run the prefix and open the Tester64.exe, start it up and then close it. After this, whenever launching the game, the camera will now turn on! But it wasn't working in-game yet, it would just turn on.


To fix that, there were a couple of other issues to solve: 
- One of them, was that the TrackIR dummy program was telling, at game startup, that it could not find the mfc42u.dll and TIRViews.dll. That was when I noticed that the symbolic links for those DLL in $HOME/.steam/steam/steamapps/compatdata/<game_id>/pfx/drive_c/Program Files (x86)/Linuxtrack were not properly linked. The links searched for the .dll in .config/linuxtrack/tir_firmware. Since I did some of the manual installing (because the installers in the tool were not working), it probably did not copy the desired .dll files in there. So I manually copied them to .config/linuxtrack/tir_firmware. 
They were found in:
`$HOME/.steam/steam/steamapps/compatdata/<game_id>/pfx/drive_c/Program Files (x86)/TrackIR5/TIRViews.dll`
and 
`$HOME/.steam/steam/steamapps/compatdata/<game_id>/pfx/drive_c/windows/syswow64/mfc42u.dll`

- Phew! This might be it, right? No! There was another issue now. 
When I did the ./ltr_extractor --update, it set the IL-2 value in $HOME/.config/linuxtrack/tir_firmware/gamedata.txt to ((null)). This should not be null at all. As much as tried calling the ./ltr_extractor --update again, it would never set the desired value to be different than ((null)). The solution was mentioned in another comment in this issue, this time by @immakyle.
Since IL-2 is an old game, you can get the old universal package with the bins already built and call the ./ltr_extractor --update present in there. Afther this, checking the gamedata.txt showed a different value!
With all of this done, I booted up IL-2 again and voila! I had a working headtracker with TrackIR on a Steam/Proton game.

Once again, a big thanks to everyone who contributed to this issue especially to @fwfa123 and @immakyle for providing such great solutions.

**Comment by tracking-enjoyer** [comment]
Hi all, also chiming in with more questions about getting linuxtrack working! Here's my setup:

- Ubuntu 24.04 with the standard DE running under Wayland
- Steam via the official `.deb`s 
- `wine-10.0`, installed through steam
- TrackIR 5, and trying to work with NP's firmware 5.4.2 from the S3 link in this project
- Using latest `master` (205f8d65107059061645f9c1a3c7a2c5828383d5) from the `uglyDwarf` fork at https://github.com/uglyDwarf/linuxtrack

Good news first -- I've gotten the project to build smoothly using the steps described in the comments above, and the software starts consistently. Going into the camera and 3D views, I can move the trackclip around in space, and it's reported correctly by both views.

Here's where things get tricky: I'm trying to install `Linuxtrack-Wine` through the `Misc.` tab in `ltr_gui`, but it consistently fails with this log message if I try to use `Download` or `Extract from installer` with a path to the `exe` on my local machine:

```
[Sun 2025-06-29 17:06:50 EDT] ===============================[Sun 2025-06-29 17:06:50 EDT] Couldn't extract poem1.txt![Sun 2025-06-29 17:06:50 EDT] Couldn't extract tir4.fw![Sun 2025-06-29 17:06:50 EDT] Couldn't extract tir5.fw![Sun 2025-06-29 17:06:50 EDT] Couldn't extract poem2.txt![Sun 2025-06-29 17:06:50 EDT] Couldn't extract sn4.fw![Sun 2025-06-29 17:06:50 EDT] Couldn't extract tir5v2.fw![Sun 2025-06-29 17:06:50 EDT] Couldn't extract game data![Sun 2025-06-29 17:06:50 EDT] Couldn't extract TIRViews.dll!
```

... It then tells me to install the firmware, which contradicts the `Firmware not needed!` field description under `Device Setup -> TrackIR Setup`.

My initial instinct was to spin up a win10 VM, try and install the firmware there, and just port over the DLLs, `.fw` files, and other files over to my Linux host using a shared volume, but after both installing the TrackIR firmware as well as pulling apart the executables, the only file I was actually able to find was `TIRViews.dll`, which now lives in the `~/.local/bin/trackir/firmware` dir on my Linux host alongside the binaries (which are in `~/.local/bin/trackir/bin`).

I know this was a janky approach to trying to solve the issue, but has anyone else seen this before? Where do those missing files come from? Am I misunderstanding the actual install flow here -- and specifically, is this something that comes strictly before doing `protontricks` stuff, or is it fixed by doing the `protontricks` steps described in some of the other comments?

Thanks!

**Comment by Goosoe** [comment]
Hello @tracking-enjoyer, hopefully this can help you out.

> Here's where things get tricky: I'm trying to install Linuxtrack-Wine through the Misc. tab in ltr_gui, but it consistently fails with this log message if I try to use Download or Extract from installer with a path to the exe on my local machine:

I wasn't able to install the linuxtrack-wine at all through the Misc. tab. I had to install it manually in the desired prefix. The .exe could be found in /share/linuxtrack/ directory. If you want to setup multiple games, you have to set it up in all the game prefixes you desire. The objective is to have the bridge installed in the prefix, not to get the .dll's. The .dll's are required to make the bridge work.

> ... It then tells me to install the firmware, which contradicts the Firmware not needed! field description under Device Setup -> TrackIR Setup.

The firmware is not needed for the "Linux part" of the installation, but needed for the wine prefix. Maybe [this can help you](https://linuxconfig.org/using-wine-prefixes) out understanding how they work! 

Protontricks is just a helpful tool to work with Steam Proton prefixes. If you want to install linuxtrack in a Steam game, for me, it is the best tool. You can call console commands and open a gui page where you can select a steam game prefix and work on it.
In the selected prefix you will have to install the TrackIR software, mfc42 package (for the .dll required fo TrackIR) and the linuxtrack-wine, which will install the bridge between the prefix environment and the "Linux part". In my post above I explain how I did it. The steam-id is a number, not the game name.
Good luck!

**Comment by tracking-enjoyer** [comment]
@Goosoe thanks, that's super helpful! I'm still missing some of the core concepts with gaming on Linux, and that link goes a long way in filling those gaps.

**Comment by fwfa123** [comment]
@Goosoe Glad it helped and thank you for the extended information. All very helpful to us die hard trackir linux gamers
@tracking-enjoyer I have been working on an ambitious project to bring linuxtrack to the newer distro versions. One issue I ran into when testing my new QT5 build on Kubuntu 25.10 was that would not runn corretly under Wayland.  You may run into this issue so you will need to run `QT_QPA_PLATFORM=xcb ltr_gui` instead of running ltr_gui directly.
I am in over 1 month of the upgrade process but I have not pushed it to a new repo (fork) yet.   But I will soon and make sure that @uglyDwarf gets all the real credit. I hope to actually make it work and I do currently have a working system on Fedora 42 and Kubuntu 25.10. Just working on the build issues at this point concerning winegcc or minGW and then I should have a basic build I can post.

**Comment by Goosoe** [comment]
@fwfa123 Thank you for your dedication on updating it! 
If you need someone to test the fork on Arch Linux, feel free to reach out! I would love to have this great piece of Software as easily accessible as possible.

**Comment by tracking-enjoyer** [comment]
@Goosoe just getting back to this now (sorry, busy week) -- I've got two followups to your original comment...

>  Finally I manually installed the previously built linuxtrack-wine.exe.

Where did you install that in the wine prefix? Just dropped straight into `drive_c/Program Files/`?

> The solution was found mentioned in a comment above: Run the prefix and open the Tester64.exe, start it up and then close it.

Just to make sure, the steps here would be...

1. Launch the app from steam for example
2. Launch `Tester64.exe` and close it

Or does "run the prefix" here just mean opening up the prefix dir?

**Comment by Goosoe** [comment]
@tracking-enjoyer Hello again. Let's see if we can get you sorted.

By "manually installing" I meant, opening the prefix through the console or using Winetricks/Protontricks( Winetricks is for any of the generic wine prefixes and Protontrics is specific for Steam Proton). I recommend using the "tricks" way because they are more user-friendly. 
You can install this package in your OS: https://ubuntu.pkgs.org/24.04/ubuntu-multiverse-arm64/protontricks_1.10.5-1_all.deb.html.
If you noticed, the linuxtrack-wine.exe has an .exe extension. Linux does not know how to work with it. If you just dropped the .exe or try to run directly, nothing useful would happen.
When you "open" the prefix through Wine, you can run .exe and other windows specific files. 
Let's have a simple example of trying to open and run an .exe in a Steam Proton prefix.

-  Open Protontricks, that will open a list of prefixes (your Steam games) for you to work on.

<img width="608" height="395" alt="Image" src="https://github.com/user-attachments/assets/4acf39cb-9efb-4565-9633-fd3235d9b48c" />

- Select "Select the default wineprefix". This will put you in the root location of the chosen prefix. Imagine that you figuratively "opened the VM". As we talked before, Wine is not a VM, but you can get the gist of it. Now, everything that you do after this window, can do Windows stuff, such as running .exe files.

<img width="998" height="489" alt="Image" src="https://github.com/user-attachments/assets/c5dc6fbf-a90d-49cb-8d6c-db3fca6271a7" />

- Scroll down and select "Run an arbitrary executable (.exe/.msi.msui)" and select the linuxtrack-wine.exe. Wherever it is, the installation will install in the prefix you are on.

<img width="997" height="490" alt="Image" src="https://github.com/user-attachments/assets/332c1f7d-7188-43eb-b8a1-a4ea9deee7df" />

- Now you have it installed! Go to the path of your prefix to check if it's there.
/$HOME/.steam/steam/steamapps/compatdata/<game_id>/pfx/drive_c/Program Files (x86)/Linuxtrack

And then, you can follow the rest of the steps. Now that you understand how to interact with your prefix, the rest should go smoothly. You can install the TrackIR software and necessary DLL's using the same process or through the console (I will leave this for you to do your own research).
Good luck!

**Comment by tracking-enjoyer** [comment]
@Goosoe thanks very much! I was brain farting on installing `linuxtrack-wine.exe` -- I'm not used to the norm on Windows being installers rather than just straight executables like on Linux, so I figured it was a question of dropping it into a dir (which clearly wasn't the case 😅 )

I've got that squared away (plus, `mfc42` is installed via protontricks), but I'm stuck now at the step with `Tester64.exe`. I can't find that executable in the prefix, though I did find _just_ `Tester.exe` in the `drive_c/Program Files (x86)/Linuxtrack` dir. Running it doesn't turn on the camera, and with `ltr_gui` running outside of the Wine prefix, `Tester.exe` wasn't reporting camera positions (though I do have a working camera preview in `ltr_gui` if I choose to turn that on, and that _does_ toggle the lights on the camera on).

I did read ahead a bit and checked if it's possibly a case of the DLLs for `mfc42u.dll` and `TIRViews.dll` being broken, but running `file` and `stat` on those confirms that they're _not_ symlinks in the wine prefix (`mfc42u.dll: PE32 executable (DLL) (GUI) Intel 80386, for MS Windows, 5 sections`), so I _think_ those are fine.

I'm also running into your issue where `gamedata.txt` has a bunch of fields populated with `((null))`, but running `ltr_extractor` from the most recent universal 64-bit release is failing due to `libcrypto.so.1.0.0` missing, which I've tried fixing by adding it to my search path (linked to `libssl.so.1.0.0`, which seems to be equivalent).

Thanks again for your time!

**Comment by Goosoe** [comment]
@tracking-enjoyer 

> I've got that squared away (plus, mfc42 is installed via protontricks), but I'm stuck now at the step with Tester64.exe. I can't find that executable in the prefix, though I did find just Tester.exe in the drive_c/Program Files (x86)/Linuxtrack dir. Running it doesn't turn on the camera, and with ltr_gui running outside of the Wine prefix, Tester.exe wasn't reporting camera positions (though I do have a working camera preview in ltr_gui if I choose to turn that on, and that does toggle the lights on the camera on).

I have both Tester.exe and Tester64.exe. Maybe your prefix is running solely on 32-bit and for that, it did not install the 64-bit version. You can run the Tester.exe to the same effect I think. Open it and press the Start button, you should start getting some values in the boxes. That means that the Prefix is able to read the camera now.
If that is not happening, most likely is that you are missing something. That thing missing, I have no clue what it might be. Maybe you are missing dependencies as you mentioned below? Maybe you need to restart your PC? You can probably try and find some logs somewhere.

The symlinks should "exist in Linux" (it's not a Wine specific thing). As I posted before in a link, Wine is not an encapsulated thing of its own. It can access all your files. You can go to the .dll's properties in the wine prefix path and see if they are properly linked. If not. Fix that.

> I'm also running into your issue where gamedata.txt has a bunch of fields populated with ((null)), but running ltr_extractor from the most recent universal 64-bit release is failing due to libcrypto.so.1.0.0 missing, which I've tried fixing by adding it to my search path (linked to libssl.so.1.0.0, which seems to be equivalent).

I did not understand if that fixed the issue or not. If it's still ((null)), maybe it's because your game is more recent? Maybe you could try some older versions of the pre-built binaries as well, just to be safe. Did you try running your self-built ltr_extractor first and see if the value was ((null))?

You can also try to run the game through Steam and see if the camera turns on automatically or if you get any popup error regarding the camera.

Hope this helps.

**Comment by fwfa123** [comment]
Thanks for sticking with this — I know it’s been a bit of a slog, but your feedback has been incredibly helpful as I work through updating the code.

Good news: the build now works without needing winegcc at all. I’ve moved all binary creation over to MinGW. The TrackIR temporary install and firmware extraction now work directly from a button in the interface, which is now fully running on Qt5.

The Wayland issue is still on the table, but that’s something I’ll address later.

I’m currently adding new buttons to the main dialog that will handle firmware and mfc42u.dll extraction automatically for the .config folder. One key update: the new firmware extraction feature will require Wine Staging, but that’ll be clearly listed in the dependencies.

Once the full firmware and DLL automation is in place, I plan to add — or replace — buttons to include Proton and Wine/Lutris installers. Hopefully, this helps resolve the setup headaches we’ve all had.

Right now, a full build from scratch (minus the new buttons mentioned above) is working with all the updated dependencies, all of which are available in Debian. My plan is to finish the button work and test across Arch, Fedora, and Debian, then share a release candidate for feedback.

I haven’t published anything publicly yet since this version could break existing working setups — and I know how much effort it took to get those running. But things are progressing well, and I hope to have something ready to share by the end of the month.

In the meantime, please keep reporting any issues — they’re extremely helpful for improving the new build.

**Comment by fwfa123** [comment]
3 steps forward and one giant step back.  I am sure that uglydwarf already experienced this but I had to find out for myself as I made a big mess leaping before I looked....the winegcc to MinGW move was a short lived win. Once I had all the automation running for installs and started with the wineprefix linuxtrack-wine installs I started having issues....so I am still plugging ahead but will have to take a different approach to make installing it as easy as I can for standard users and also offer the full install using another approach for a deeper dig approach

The original issue was that `check_data.exe` built with MinGW was crashing under Wine with a null pointer dereference. After investigation, I discovered that:

1. **MinGW builds create PE32 files** that run under Wine but cannot create Unix symlinks
2. **Symlinks are essential** for LinuxTrack/Wine integration (Linux firmware → Wine DLLs)
3. **The working version used winegcc** which creates native Linux ELF binaries with symlink support
4. **32-bit compatibility** is required for most old games

## Solution Overview

I've implemented a comprehensive solution with multiple user options:

### 1. **Pre-built Binaries (Recommended for Users)**
- 32-bit ELF binaries built with winegcc
- Includes installation script with Wine prefix detection
- No development packages required for end users

### 2. **Build from Source (For Developers/Advanced Users)**
- Automatic detection of winegcc vs MinGW..I may simply drop all the MinGW code
- Distribution-specific dependency installation
- Clear documentation for all major Linux distributions

### 3. **Hybrid Code Approach**
- Single source file that works with both winegcc and MinGW...may drop..
- Graceful fallback when winegcc is unavailable
- Clear warnings about functionality limitations

Just an update to let you guys know why I have not published yet

**Comment by Cenarkion** [comment]
Watching this thread intently!  Have been bouncing around  with attempts to get this to play ball - and it feels close, but not quite there yet!
Running Nobara (fedora), the universal installer seems to deploy, mostly ok - though there are some libraries that seem to have no package to provide the missing pieces:

 ldd * | grep 'not found' | sort -u
        libcrypto.so.1.0.0 => not found
        libcwiid.so.1 => not found

Thus ltr_extractor wont play, but ltr_gui works perfectly (head tracking in the tracking window is flawless, no lag, its a wonder full tease! :-D )
I cannot seem to get it to play  nicely with DCS (running through Lutris) however - I'm probably missing something obvious - but my fumbling so far as been to no avail.

Installing Linuxtrack-Wine via the misc tab was an erroring failure.  I then ran the linuxtrack-wine.exe, via  Lutris, theoretically i nstalling into the DCS prefix - which seems to work, but I  see no change in behaviour.  I can see a drive_c/Program Files (x86)/Linuxtrack directory with a hanful of files therein, however attempting to run Tester.exe or Tester64.exe from therein presents an invalid handle message, followed by a number of File not found errors.  Possibly related to the above missing libraries, though Im unsure.

**Comment by immakyle** [comment]
@Cenarkion A couple things...I'll try and keep this concise and easy to follow:

**1. Ensure you are using linuxtrack-wine.exe from a working fork, not from the universal package (it won't function).**
 
Both these should be fine, I just built them to double-check. Pick one.
git clone https://github.com/RavenX8/linuxtrack.git
git clone https://github.com/marengohue/linuxtrack.git

Build using method provided by @fwfa123: 
I followed the instructions nearly identically with a change to the --with-wine-libs flag in the ./configure step.
My exact commands are as follows:

cd linuxtrack/
aclocal
autoupdate
autoconf
automake
./configure --prefix=$HOME/linuxtrack --with-wine-libs=-L/usr/lib/wine
make -j

***NOTE: I'm not sure about Nobara/Fedora repos, but due to the recent changes to the wine package on Arch we need to use [wine-stable](https://aur.archlinux.org/packages/wine-stable) and [wine-stable-mono](https://aur.archlinux.org/packages/wine-stable-mono) from AUR.*** This change was mentioned by @Goosoe and was not present the first time I put linuxtrack together. 

You'll know if you have the right wine package after you run "make -j" you'll see the following summary...
checking Webcam support... yes
checking Wiimote support... no
checking TrackIR support... yes
checking Facetracker support... yes
checking XPlane plugin... no
checking Mickey... yes
checking Wine plugin... yes
checking OSC support... yes
If you see a yes next to Wine plugin, it worked. If not, then you'll need the equivalent of wine-stable and wine-stable-mono on Nobara/Fedora.

sudo make install
sudo cp $HOME/linuxtrack/share/linuxtrack/99-TIR.rules /lib/udev/rules.d
reboot

The compiled binaries will be found in $HOME/linuxtrack/bin/
linuxtrack-wine.exe in $HOME/linuxtrack/share/linuxtrack/

**2. Let's now shift our focus to the universal package.**
 
I'm missing more libraries than you. This is my output:

cd /opt/linuxtrack-0.99.18/bin/
ldd * | grep 'not found' | sort -u
	libcwiid.so.1 => not found
cd ../lib/linuxtrack
ldd *.so  | grep 'not found' | sort -u
	libcwiid.so.1 => not found
	libopencv_core.so.2.3 => not found
	libopencv_imgproc.so.2.3 => not found
	libopencv_objdetect.so.2.3 => not found

None of that matters for using "ltr_extractor --update" command from the universal package, which updates your gamedata.txt found in .config/linuxtrack. 

However, you are missing libcrypto.so.1.0.0. On Arch, we can get that using [openssl-1.0](https://aur.archlinux.org/packages/openssl-1.0) found in AUR. @tracking-enjoyer you were missing this library as well, which caused ltr_extractor to fail. I don't know how to get it on your respective distros, but it is what you need. After you acquire it "ltr_extractor --update" will work. Check gamedata.txt to see that it has changed. Should add more supported game titles, and without ((null)) entries.


To summarize the purpose of these steps:
1. Build linuxtrack fork for a functioning wine-bridge (linuxtrack-wine.exe). Then install it into game prefix using your preferred method i.e. Lutris, protontricks, command line.
2. Get universal package working for gamedata.txt updates (ltr_extractor --update)

**Comment by fwfa123** [comment]
Some good news as I saw the post above.  I did update the linuxtrack-wine.exe file so it plays with new distros very well. The wine install button is working on the gui again properly, handles everything you need in one go  just as uglydwarf originally designed it. (winetricks is required to be installed, also wine development but I have a script to automate the install of this (wine development package) on most distros now)  I also updated the extractor code so now the game_data.txt file is generated without any nulls rather all the games will come out with the correct code number attached. The issue was that the new trackir installer exe had changed how they created the game file so the code to pull that data needed to be updated.  Closer and closer I am getting to a release...oh I also now Dock the tracking window as a choice for the user.  I really want to get a button for Lutris and Proton linuxtrack-wine install created as a one stop install but I am not sure I will get that before I want to publish this.... Sorry for it being drawn out but I only have time to work on it on the weekend

**Comment by Goosoe** [comment]
@fwfa123 Thank you for the updates.
You are doing a great job, taking into account you are doing it in your free time for the Linux community. I am sure we all appreciate it!
Do take your time, and don't rush. We can wait a while longer!

**Comment by fwfa123** [comment]
First one is published..lots more to do but it is ready to test for anyone how wants too....appimage approach in the releases. see the main README file for building instructions.  
[https://github.com/fwfa123/linuxtrackx-ir/releases/tag/v0.99.21](url)
Lots to be told in the TODO [https://github.com/fwfa123/linuxtrackx-ir/blob/v0.99.21/docs/progress/TODO_FUTURE_DEVELOPMENT.md](url)

**Comment by fwfa123** [comment]
Well I saw a good many issues with my Appimage under Arch that I did not catch the first time prompting me to take yet another path.  One good thing I found was that the trackir installer no longer uses the old mfc42 dll which was more difficult to get a hold of and causing some of the modernization problems.  It now uses mfc140.  I will update this and then update the appimage in the pre-release soon.  (UPDATE this was not true and I moved back to MFC42 with a requirement for the user to have Winetricks installed)

While getting ready to pass on  some more info I realized that I need to allow issues and discussions in the repo. Venturing in to these settings I saw 2FA alerts and at the same time being alerted by  in my email to Githubs requirement that I use 2FA due to the fact that I am submitting code.  I am being told that they will lock me out of the code without this...sorry to say that will be it for me.  I will have to move copy the project to another git that does not require this action.  

I have until September 9th of 2025 to comply and up and to that point I will update on my Linuxtrackx-ir version and slowly transfer to another git.  

The world is shrinking and I am one of those nuts that does not fit  lol

**Comment by fwfa123** [comment]
Updated with a new release and it will be my last one on Github moving to Gitlab for my next updates.  same user name..
I did update the UI to add installers that work for Proton and Lutris, plus regular wine.  I also added a system tab for more info in one place when trying to debug
<img width="1418" height="691" alt="Image" src="https://github.com/user-attachments/assets/d9ebce54-a6df-44bb-ab26-de01b86d28cc" />

**Comment by Goosoe** [comment]
Amazing work @fwfa123!
Maybe you could put the note about moving to Gitlab in your readme file as a header so it's not that easy to miss?
Thank you!

**Comment by fwfa123** [comment]
Good idea, will do Thanks!
